[25.0] More fixes to `FormData` drag and drop and typing

ahmedhamidawan opened the pull request:
@mvdbeek has pushed a commit that should fix map-over of collection DCEs. Fixes https://github.com/galaxyproject/galaxy/issues/20315

The rest of these changes (from a while ago 😅) are improvements/fixes for drag and drop as a follow up to changes in https://github.com/galaxyproject/galaxy/pull/19866.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
type the return of `createDatasetCollection` as `HDCASummary`

ahmedhamidawan committed:
improve alert handling in `FormData`, and `collectionTypeToText`

ahmedhamidawan committed:
fix `keepOptions` to actually trigger an update on `formattedOptions`

For some reason, I had to use a counter `ref` to trigger this update as without it (a computed on `keepOptions`, turning `keepOptions` itself to a `ref` etc) did not update when `keepOptions` was changed.

ahmedhamidawan committed:
get `collectionType` and `name` for `DCObject`s as well

ahmedhamidawan committed:
disallow dragging collections to dataset-only inputs

This refers to `FormData` elements without the collection variant (only single or multiple dataset variants).

Explained in https://github.com/galaxyproject/galaxy/pull/19866#issuecomment-2744061385

ahmedhamidawan committed:
jest fix for changed error based on `COLLECTION_TYPE_TO_LABEL`

ahmedhamidawan committed:
fix jest since we removed `$emit("alert", undefined);` from `onDrop` function

mvdbeek commented:
We can’t do that, this breaks reruns of mapped over jobs.

mvdbeek commented:
That’s not what we discussed in the backend meeting, you can’t allow paired collections in a non-paired input is the issue

ahmedhamidawan commented:
> That’s not what we discussed in the backend meeting, you can’t allow paired collections in a non-paired input is the issue

But then, if you try a (non DCE) pair as an input here for this, it doesn't error out.

mvdbeek commented:
I am not sure I understand this comment ? A DCE representing a list should and does work currently.

ahmedhamidawan commented:
Sorry, since you said:

>  you can’t allow paired collections in a non-paired input is the issue

I tried using a pair as in input to this same tool, and it doesn't produce the error (from Jen's issue). Its only when I try a pair that is inside a collection that this fails.

mvdbeek commented:
We're talking about DCEs, so that's what I'm doing. And it works on lists inside of list:list collections.

ahmedhamidawan commented:
Ah ok

So it's a particular thing where, on the client we can look for if it's a DCE that is a paired collection, disallow it as a drag and drop input unless that input specifically allows that

ahmedhamidawan commented:
> We're talking about DCEs, so that's what I'm doing. And it works on lists inside of list:list collections.

I just tried `list:list` and it produces that error too:

https://github.com/user-attachments/assets/fdb87e56-96fd-4cb4-a1c1-b966a62f9090

mvdbeek committed:
Fix/Enable drag-and-drop of collection DCEs on multiple="false" data parameters

We might want to consider creating a hidden HDCA from the dragged DCE(s) first
though ?

mvdbeek commented:
I've pushed a commit that should fix map-over of collection dces.

ahmedhamidawan commented:
Oh nice, so this fixes the bug reported in https://github.com/galaxyproject/galaxy/issues/20315 as well, because now we successfully accept DCE collections (even pairs!) as valid inputs here:

https://github.com/user-attachments/assets/c4b6d440-8046-485a-82f3-497091df0475

If a `FormData` has explicit collection types, that is already something we check and confirm against even on drag and drop.

So, I will remove the commit now that disallowed DCE collections as inputs (and update the PR description...)

ahmedhamidawan commented:
Good point, I can add jests as well as backend tests.

ahmedhamidawan committed:
add API test for DCEs being acceptable as mapped over inputs on drop
